src/server/pfs/fuse/server.go

```diff
@@ -1210,21 +1229,120 @@ func sanitizeInputAndGetAlias(datumInput *pps.Input, c *client.APIClient) (map[s
 		} else {
 			datumInputsToMounts[pfsInput] = []string{input.Pfs.Name}
 		}
-
 		return nil
 	}); err != nil {
+		return true, err
+	}
+	func() {
+		mm.mu.Lock()
+		defer mm.mu.Unlock()
+		mm.DatumInput = datumInput
+		mm.DatumInputsToMounts = datumInputsToMounts
+	}()
+	return useGlobFile, nil
+}
+
+func (mm *MountManager) GetNextXDatums(numDatums int, paginationMarker string) ([]*pps.DatumInfo, error) {
+	req := &pps.ListDatumRequest{
+		Input:            mm.DatumInput,
+		Number:           int64(numDatums),
+		PaginationMarker: paginationMarker,
+	}
+
+	ctx, cf := context.WithCancel(mm.Client.Ctx())
+	defer cf()
+	client, err := mm.Client.PpsAPIClient.ListDatum(ctx, req)
+	if err != nil {
 		return nil, err
 	}
+	var dis []*pps.DatumInfo
+	for {
+		di, err := client.Recv()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				return dis, nil
+			}
+			return nil, err
+		}
+		dis = append(dis, di)
+	}
+}
 
-	return datumInputsToMounts, nil
+func (mm *MountManager) GetDatums(useGlobFile bool) (retErr error) {
+	defer func() {
+		retErr = grpcutil.ScrubGRPC(retErr)
+	}()
+	// TODO: Try creating datums using GlobFile for now because it's faster than ListDatum.
+	// Once ListDatum pagination is efficient, we can remove this. #ListDatumPagination
+	if useGlobFile {
+		return mm.CreateDatums()
+	}
+	// If input spec has Join or Group, fall back to ListDatum
+	datums, err := mm.GetNextXDatums(NumDatumsPerPage, "")
+	if err != nil {
+		return err
+	}
+	if len(datums) == 0 {
+		return errors.New("no datums to mount")
+	}
+	func() {
+		mm.mu.Lock()
+		defer mm.mu.Unlock()
+		mm.Datums = datums
+		mm.PaginationMarker = datums[len(datums)-1].Datum.Id
+		if len(datums) < NumDatumsPerPage || NumDatumsPerPage == 0 {
+			mm.AllDatumsReceived = true
+		}
+	}()
+	return nil
 }
 
-func getCopyOfMapping(datumInputsToMounts map[string][]string) map[string][]string {
-	datumInputsToMountsCopy := map[string][]string{}
-	for k, v := range datumInputsToMounts {
-		datumInputsToMountsCopy[k] = v
+func (mm *MountManager) CreateDatums() error {
```

Comment:
i think it may be a good idea to document the algorithm a bit, it's not very straightforward to understand just reading through the code. though that may just be because i'm not used to some of go's semantics.